impress-org/givewp
# Add PayPal Donations Payment Gateway

Resolves #4886 

## Description

This PR merges the entire PayPal Commerce epic and is intended to be **the last** thing merged when all epic issues have been closed. No funny business. I'm watching you.

Threats aside, the intention is to be able to test the Epic and provide feedback. Please see the testing instructions below.

### Terminology

There are a few terms helpful to know.

**PPCP & Express Accounts**
PPCP accounts are accounts that have been authorized by PayPal to have people supply credit cards directly from their website. This requires extra authorization and is limited to 28 countries (mostly Americas and Europe). All other countries must use Express accounts. These accounts are authorized to accept payments, but not accept credit cards directly. In short, PPCP accounts can use Advanced Fields (see below) and Smart Buttons, while Express accounts can only use Smart Buttons.

**Smart Buttons**
When viewing the form, you'll see some large, bold colored buttons. These are [Smart Buttons](https://developer.paypal.com/docs/checkout/#try-the-buttons). Smart Buttons are clever buttons that PayPal provides for processing payments, and are tightly regulated by PayPal. Most buttons take the donor off-site, briefly, except the "Debit or Credit Card" button, but is a special (but hand-cuffed) button that only shows up when Advanced Fields aren't present.

**Advanced Fields**
When using a PPCP account, the fields below "or pay with card" line are Advanced Fields. These are fields that feel like perfectly normal form fields. They do special things on the Seller's account, and provide the simplest donor experience. They can only be used on PPCP accounts, however.

## Visuals

### Testing a PPCP account


## Pre-review Checklist



-   [ ] All epic issues are resolved and closed

## Testing Instructions

In general:
- After pulling the branch, run `composer install && npm i && npm run dev` on your CLI
- Make sure, when connecting accounts and making donations, you are in test mode.
- Use `4242 4242 4242 4242` as your test credit card number with a future date expiration and any 3-digit CVC number

### Express account donation
1. Go to the GiveWP General settings and set the country to Columbia (note: Do not change the currency to Colombian Peso as it's not supported by PayPal, just use USD or Euro)
2. Go to the PayPal Donations connection screen
3. Connect using the following credentials:
    - Email: sb-siizv2863201@business.example.com
    - Pass: &;9jU+Y5
4. Make a donation and subscription using Smart Buttons


### PPCP US account donation
1. Go to the GiveWP General settings and set the country to United States
2. Go to the PayPal Donations connection screen
3. Connect using the following credentials:
    - Email: sb-kttpp2709027@business.example.com
    - Pass: 3lU<#qG.
4. Make a donation and subscription using the Smart Buttons
5. Make a donation and subscription using the CC fields

### PPCP French account donation
1. Go to the GiveWP General settings and set the country to France
2. Go to the PayPal Donations connection screen
3. Connect using the following credentials:
    - Email: sb-i8fua2524394@business.example.com
    - Pass: Qf>a^9CM
4. Make a donation and subscription using the Smart Buttons
5. Make a donation using the CC fields
6. The CC fields should not show up when recurring is enabled (only US and AU are supported at this time)

### Refunding
1. Go to a donation made with with this gateway in the admin
2. Change the status to Refunded
3. Save the donation
4. Click on the transaction id and see if the payment has been refunded
![image](https://user-images.githubusercontent.com/2024145/90079045-a432a900-dcbb-11ea-9d76-d236b635742b.png)
